Create `@shopify/plugin-polaris` package

sam-b-rose opened the pull request:
<!--
  ☝️How to write a good PR title:
  - Prefix it with [ComponentName] (if applicable), for example: [Button]
  - Start with a verb, for example: Add, Delete, Improve, Fix…
  - Give as much context as necessary and as little as possible
  - Prefix it with [WIP] while it’s a work in progress
-->

### WHY are these changes introduced?

Related #6614 

[This previously lived](https://github.com/Shopify/cli/pull/358) in the [Shopify/cli](https://github.com/Shopify/cli/) repo, but has been moved to the Polaris repo for better maintainability and colocation.

<!--
  Context about the problem that’s being addressed.
-->

### WHAT is this pull request doing?

Adds a Polaris CLI package for integrating with the Shopify CLI as a plugin.

#### Plugin usage

```sh
# install the plugins
shopify plugin install polaris

# list available commands
shopify polaris help

# use the migration tool
shopify polaris migrate <migration> <path>
```

sam-b-rose committed:
Create @shopify/polaris-cli package

aaronccasanova commented on polaris-cli/tsconfig.json:
```diff
@@ -0,0 +1,12 @@
+  "extends": "@shopify/typescript-configs/node-base",
+  "compilerOptions": {
+    "module": "commonjs",
+    "target": "es2019",
```
```suggestion
    "target": "es2020",
```

Suggest bumping to `es2020` for Node.js v14 support (note: This is the [minimum Node.js version required by the Shopify CLI](https://shopify.dev/apps/tools/cli#requirements))

aaronccasanova commented on polaris-cli/package.json:
```diff
@@ -0,0 +1,60 @@
+  "files": [
+    "/bin",
+    "/dist"
+  ],
```
```suggestion
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "bin": "bin/run",
  "files": [
    "bin",
    "dist"
  ],
```

Light suggestion for consistency

pepicrft commented on polaris-cli/README.md:
````diff
@@ -0,0 +1,41 @@
+You can install the Polaris plugin using the Shopify CLI by running the following command:
+
+```sh
+shopify plugins install @shopify/polaris-cli
````
The pattern we've followed for Hydrogen is calling the package `cli-xx`. Unless there's a reason to make this one `polaris-cli`, I'd suggest following that pattern.

sam-b-rose replied:
I've updated it to be `@shopify/plugin-polaris`. If I understand the [friendly names](https://github.com/oclif/plugin-plugins#friendly-names) correctly, this should allow users to install in with `shopify plugins install polaris`

pepicrft commented on polaris-cli/bin/dev:
```diff
@@ -0,0 +1,16 @@
+const path = require('path');
```
If you follow the suggestions that I gave you, I don't think you need any executable in this repository besides the one that comes with the `@shopify/cli` package.

pepicrft commented on polaris-cli/src/commands/polaris/migrate.ts:
```diff
@@ -0,0 +1,43 @@
+import {Command, Flags} from '@oclif/core';
+import {run as runMigrator} from '@shopify/polaris-migrator';
+
+export default class Migrate extends Command {
```
I'd recommend inheriting from the class exported by `@shopify/cli-kit`. [Here's](https://github.com/Shopify/cli/blob/main/packages/app/src/cli/commands/app/dev.ts#L6) an example.

sam-b-rose replied:
Looks like it's having a hard time finding the types for this import. Any suggestions?

<img width="1144" alt="Screen Shot 2022-09-14 at 12 48 16 PM" src="https://user-images.githubusercontent.com/11774595/190238029-811ac22c-2627-467e-97b4-738a92130902.png">

pepicrft commented on polaris-cli/tsconfig.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "extends": "@shopify/typescript-configs/node-base",
+  "compilerOptions": {
+    "module": "commonjs",
```
Are you planning to support old versions or node or do you have a dependency that's not consumable from ESM? If not, I'd recommend [aligning with the CLI ](https://github.com/Shopify/cli/blob/main/configurations/tsconfig.json#L5) and use ESM

sam-b-rose committed:
Keep migrator command and CLI in sync

This PR also removes the ability to use the Polaris CLI as a standalone tool. It is now only used as a plugin with Shopify CLI.

sam-b-rose committed:
Fix --help message formatting for migrator CLI

aaronccasanova committed:
Polaris CLI experiments (#7187)

Co-authored-by: Sam Rose <sam.rose@shopify.com>

sam-b-rose committed:
Restructure migrator types and file structure
